File: base.mvp/src/main/java/alvin/base/mvp/main/MainModule.java
package alvin.adv.mvp.main;

import alvin.adv.mvp.main.presenter.MainPresenter;
import alvin.adv.mvp.main.views.MainActivity;
import dagger.Binds;
import dagger.Module;
import dagger.android.ContributesAndroidInjector;

@Module
public interface MainModule {

    @ContributesAndroidInjector(modules = {
            ViewModule.class,
            ProviderModule.class
    })
    MainActivity mainActivity();

    @Module
    interface ViewModule {

        @Binds
        MainContracts.View view(MainActivity activity);

        @Binds
        MainContracts.Presenter presenter(MainPresenter presenter);
    }

    @Module
    class ProviderModule {
    }
}

File: adv.camera/src/main/java/alvin/adv/camera/main/MainModule.kt
package alvin.adv.camera.main

import alvin.adv.camera.main.presenters.MainPresenter
import alvin.adv.camera.main.views.MainActivity
import dagger.Binds
import dagger.Module
import dagger.android.ContributesAndroidInjector

@Module
interface MainModule {

    @ContributesAndroidInjector(modules = [ViewModule::class])
    fun mainActivity(): MainActivity

    @Module
    interface ViewModule {

        @Binds
        fun view(activity: MainActivity): MainContracts.View

        @Binds
        fun presenter(presenter: MainPresenter): MainContracts.Presenter
    }
}
